rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,335 @@
+executing. In that case the worker MUST top running the job.
+
+#### Mutual Exclusion
+
+If each worker updates their job status on time, and the activity monitor is
+running correctly, the same job SHOULD be executed by no more than one worker
+at a time. But, if the worker process is blocked for too long (for instance, in
+an overload scenario), it may fail to update its status often enough, and the
+activity monitor MAY re-enqueue the job. However, even in such cases, the
+mutual exclusion constraint can be maintained as long as `update(Tx,...)`
+function is called in the same transaction as the type-specific DB writes, and
+its result is checked for the `worker_conflict` return.
+
+# Implementation Details
+
+## Data Model
+
+`("couch_jobs", "data", Type, JobId) = (Sequence, WorkerLockId, JobOpts)`
+`("couch_jobs", "pending", Type, Priority, JobId) = ""`
+`("couch_jobs", "watches", Type) = Sequence`
+`("couch_jobs", "activity_timeout", Type) = ActivityTimeout`
+`("couch_jobs", "activity", Type, Sequence) = JobId`
+
+`Type`, `Priority`, and `JobId` MUST be any type supported by the FoundationDB
+tuple layer.
+
+`JobOpts` MUST be a JSON object encoded as a string. It MAY have `"priority"`,
+`"cancel"`, `"resubmit"`, or `"data"` fields.
+
+## Job Lifecyle Implementation
+
+This section describes how the framework implements some of the API functions.
+
+`add()` :
+  * Add the new job to the main jobs table
+  * If a job with the same `JobId` exists, return an error
+  * `WorkerLockId` is set to `null`.
+
+`accept()` :
+ * Generate a unique `WorkerLockId` UUID.
+ * Attempt to dequeue the item from the pending queue, then assign it the
+   `WorkerLockId` in the jobs table.
+ * Create entry in the `"activity"` subspace.
+ * Update sequence in the `"watches"` subspace for that job's type.
+
+`update()`:
+ * Check if `WorkerLockId` matches, otherwise return `worker_conflict`
+ * Check if `"canceled"` field in the `JobOpts` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
+ * Update `JobOpts`
+ * Create a new `"activity"` sequence entry and in main job table
+ * Update `"watches"` sequence for that job type
+
+`finish()`:
+ * Check if `WorkerLockId` matches, otherwise returns `worker_conflict`
+ * Check if `Canceled` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
+ * If `"resubmit"` field in `JobOpts` is `true`, re-enqueue the job into the pending queue.
+ * Set job table's `WorkerLockId` to `null`
+
+## Activity Monitor Implementation
+
+There is an activity monitor running for each job type. It performs periodic
+scans of the `"active"` range with sequences less than `"watches"` sequence for
+that type. The scan is performed as often as `ActivityTimeout`. The first time
+it scans the entries, it remembers the set of `JobId` and sequences it found.
+Subsequently, if those haven't changed since the last `ActivityTimeout` period,
+it will re-enqueue those jobs into the pending queue.
```

Comment:
I think you might have to elaborate a bit more on the Activity monitor. What happens if the activity monitor crashes and forgets the sequence number it last checked? Is it one activity monitor per node or per job type or just one.

Reply:
Good call. I'll update the activity monitor section to both expand on how the algorithm work and list the steps separately as a bulleted list.